tensorflow_datasets/core/utils/tf_utils.py

```diff
@@ -135,6 +136,7 @@ def is_dtype(value):
   return True
 
 
+@functools.lru_cache
```

Comment:
Datasets can have GB or TB of examples, this basically force the full dataset to be kept in memory.

Reply:
In fact, this function has very little correlation with the dataset size. It is directly related to the number of features. 
If we want to be on the safe side, we can limit the cache size.

Reply:
Unless each element can have a different shape. Is it possible?

Reply:
I came across this commit accidentally. Please don't assume the number of features are constant throughout the dataset. This assumption would be incorrect in the case of graph datasets, such as [ogbg_molpcba](https://www.tensorflow.org/datasets/catalog/ogbg_molpcba) on TFDS.